physics/discrete_trajectory2.hpp

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <list>
+#include <memory>
+
+#include "absl/container/btree_map.h"
+#include "base/not_null.hpp"
+#include "geometry/named_quantities.hpp"
+#include "physics/degrees_of_freedom.hpp"
+#include "physics/discrete_trajectory_iterator.hpp"
+#include "physics/discrete_trajectory_segment_iterator.hpp"
+#include "physics/discrete_trajectory_types.hpp"
+#include "physics/trajectory.hpp"
+#include "serialization/physics.pb.h"
+
+namespace principia {
+namespace physics {
+
+template<typename Frame>
+class DiscreteTrajectorySegment;
+
+namespace internal_discrete_trajectory {
+
+using base::not_null;
+using geometry::Instant;
+using geometry::Position;
+using geometry::Velocity;
+using physics::DegreesOfFreedom;
+
+template<typename Frame>
+class DiscreteTrajectory2 : public Trajectory<Frame> {
+ public:
+  using Iterator = DiscreteTrajectoryIterator<Frame>;
+  using SegmentIterator = DiscreteTrajectorySegmentIterator<Frame>;
+
+  DiscreteTrajectory2() = default;
+
+  // Moveable.
+  DiscreteTrajectory2(DiscreteTrajectory2&&) = default;
+  DiscreteTrajectory2& operator=(DiscreteTrajectory2&&) = default;
+  DiscreteTrajectory2(const DiscreteTrajectory2&) = delete;
+  DiscreteTrajectory2& operator=(const DiscreteTrajectory2&) = delete;
+
+    Iterator begin() const;
+  Iterator end() const;
+
+  Iterator rbegin() const;
+  Iterator rend() const;
+
+  Iterator find(Instant const& t) const;
+
+  Iterator lower_bound(Instant const& t) const;
+  Iterator upper_bound(Instant const& t) const;
+
+  SegmentIterator segments_begin() const;
+  SegmentIterator segments_end() const;
+
+  SegmentIterator segments_rbegin() const;
+  SegmentIterator segments_rend() const;
```

Comment:
Same comment as for rbegin, `std::reverse_iterator<SegmentIterator>`.

Reply:
Done.